examples/mixed_signal/tests/test_regulator.py

```diff
@@ -0,0 +1,242 @@
+import cocotb
+from cocotb.regression import TestFactory
+from cocotb.triggers import Timer
+from cocotb.utils import get_sim_time
+
+from collections import namedtuple
+
+import datetime
+import itertools
+import logging
+import matplotlib
+import matplotlib.pyplot as plt
+
+
+Dataset = namedtuple('Dataset', 'time, trim, voltage, current')
+
+class Regulator_TB(object):
+    """Class for collecting testbench objects.
+
+    Args:
+        dut: The toplevel of the design-under-test.
+            In this mixed cocotb/HDL testbench environment, it is the HDL testbench.
+        settling_time_ns (int): Time in nanoseconds to wait before sample is taken.
+    """
+
+    def __init__(self, dut, settling_time_ns=1):
+        self.dut = dut
+        self.settling_time_ns = settling_time_ns
+        self.analog_probe = dut.i_analog_probe  #: The instance name of the analog probe module.
+        self._bittogglestream = itertools.cycle(range(2))  #: Produce stream that toggles between ``0`` and ``1``.
+    
+    def probe_values(self):
+        """Collect data from instance pointed to by :attr:`analog_probe`."""
+        voltage = self.analog_probe.voltage.value
+        current = self.analog_probe.current.value
+        return Dataset(time=get_sim_time(units='ns'),
+                       trim=self.dut.trim_val.value.signed_integer,
+                       voltage=voltage,
+                       current=current)
+    
+    @cocotb.coroutine
+    def _get_single_sample(self, node):
+        """Measure a single voltage/current pair on *node*."""
+        toggle = next(self._bittogglestream)
+        self.dut.i_analog_probe.node_to_probe <= node
+        self.analog_probe.probe_voltage_toggle <= toggle
+        self.analog_probe.probe_current_toggle <= toggle
+        yield Timer(5, units='ps')  # NOTE: needs some time for some reason
```

Comment:
Does `yield NextTimestep()` or `yield ReadWrite()` work here instead?

Reply:
IIRC I experimented with that and didn't succeed, I'll try again.

Reply:
It's a can of worms. 
``test_rescap``:
* works when I use ``yield ReadOnly()``.
* with ``ReadWrite()``, errors out with ``100000.00ns ERROR    cocotb.scheduler                            __init__.py:191  in _sim_event                      Failing test at simulator request before test run completion: Simulator shutdown prematurely`` (no error before, but no analog value sampling either)
* with ``NextTimeStep``, the time difference where it samples the analog values first doubles every time, then samples every 2000 ns, so the testcase does not finish until the analog simulator's kernel stop time ends the simulation.